Dockerfile

```diff
@@ -1,45 +1,86 @@
+  <<HEREDOC
+    # https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/reference.md#example-cache-apt-packages
+    # https://stackoverflow.com/questions/66808788/docker-can-you-cache-apt-get-package-installs#comment135104889_72851168
+    rm -f /etc/apt/apt.conf.d/docker-clean
+    echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache
+
+    apt update && apt install -y --no-install-recommends \
+      build-essential \
+      curl \
+      python3-venv \
+      cmake
+HEREDOC
+
 ENV HOME="/root"
+ENV PATH="$HOME/.venv/bin:$PATH"
 WORKDIR $HOME
 
-RUN apt update \
-  && apt install -y --no-install-recommends \
-  build-essential \
-  curl \
-  python3-venv \
-  cmake \
-  && apt clean \
-  && rm -rf /var/lib/apt/lists/*
-
-# Setup zig as cross compiling linker
-RUN python3 -m venv $HOME/.venv
-RUN .venv/bin/pip install cargo-zigbuild
-ENV PATH="$HOME/.venv/bin:$PATH"
+# Setup zig as cross compiling linker:
+RUN <<HEREDOC
+  python3 -m venv $HOME/.venv
+  .venv/bin/pip install --no-cache-dir cargo-zigbuild
+HEREDOC
 
-# Install rust
-ARG TARGETPLATFORM
-RUN case "$TARGETPLATFORM" in \
-  "linux/arm64") echo "aarch64-unknown-linux-musl" > rust_target.txt ;; \
-  "linux/amd64") echo "x86_64-unknown-linux-musl" > rust_target.txt ;; \
-  *) exit 1 ;; \
-  esac
-# Update rustup whenever we bump the rust version
-COPY rust-toolchain.toml rust-toolchain.toml
-RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y --target $(cat rust_target.txt) --profile minimal --default-toolchain none
+# Install rust:
 ENV PATH="$HOME/.cargo/bin:$PATH"
-# Installs the correct toolchain version from rust-toolchain.toml and then the musl target
-RUN rustup target add $(cat rust_target.txt)
-
-# Build
-COPY crates crates
-COPY ./Cargo.toml Cargo.toml
-COPY ./Cargo.lock Cargo.lock
-RUN cargo zigbuild --bin uv --target $(cat rust_target.txt) --release
-RUN cp target/$(cat rust_target.txt)/release/uv /uv
-# TODO(konsti): Optimize binary size, with a version that also works when cross compiling
-# RUN strip --strip-all /uv
+COPY rust-toolchain.toml .
+RUN <<HEREDOC
+  # Install rustup, but skip installing a default toolchain as we only want the version from `rust-toolchain.toml`:
+  curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y --profile minimal --default-toolchain none
+
+  # When rustup installs the toolchain ensure it actually uses the minimal profile, avoiding excess layer weight:
+  # https://github.com/rust-lang/rustup/issues/3805#issuecomment-2094066914
+  echo 'profile = "minimal"' >> rust-toolchain.toml
+  echo 'targets = [ "aarch64-unknown-linux-musl", "x86_64-unknown-linux-musl" ]' >> rust-toolchain.toml
+  # Add the relevant musl target triples (for a building binary with static linking):
+  # Workaround until `ensure` arrives: https://github.com/rust-lang/rustup/issues/2686#issuecomment-788825744
+  rustup show
+HEREDOC
```

Comment:
When installing `rustup` in the first line, as the commit history notes (and the PR description), setting `--target` was redundant, it emits a warning that it will ignore it.

Likewise there was a mishap with the `--profile minimal` being ignored when `rust-toolchain.toml` exists and doesn't define it's own `profile` key. If you'd like that file could add this line instead? I've raised an issue upstream to get the regression in `rustup` fixed.

The next line added to `rust-toolchain.toml` is probably not something you'd want to add to the actual file, since both targets aren't mandatory 🤷‍♂️ 
- I could instead add them individually with `rustup target add ...` in their respective stages that come afterwards
- Due to the `--default-toolchain none` this still requires the `rustup show` workaround, otherwise you'd install the toolchain twice. An alternative is to use an ARG directive for passing in the toolchain version to use for `--default-toolchain`, that way you could just build with the current stable release, but if you really need to install a toolchain that matches `rust-toolchain.toml`, provide that via the ARG instead?